lib/ecto/embedded.ex

```diff
@@ -64,39 +63,109 @@ defmodule Ecto.Embedded do
   @doc """
   Wraps embedded models in changesets.
   """
-  def change(%Embedded{cardinality: :one}, value) do
-    Ecto.Changeset.change(value)
+  def change(%Embedded{cardinality: :one}, value, current) do
+    do_change(value, current)
   end
 
-  def change(%Embedded{cardinality: :many, container: :array}, value) do
-    Enum.map(value, &Ecto.Changeset.change/1)
+  def change(%Embedded{cardinality: :many, container: :array, embed: mod},
+             value, current) when is_list(value) do
+    {pk, _} = primary_key(mod)
+    current = process_current(current, pk)
+    map_changes(value, pk, &do_change/2, current, [], true)
+    |> elem(1) # It should never fail here
+  end
+
+  defp do_change(value, nil) do
+    %{Changeset.change(value) | action: :insert}
+  end
+
+  defp do_change(nil, current) do
+    %{Changeset.change(current) | action: :delete}
+  end
+
+  defp do_change(%Changeset{model: current} = changeset, current) do
+    %{changeset | action: :update}
+  end
+
+  defp do_change(%Changeset{}, _current) do
+    raise ArgumentError, "embedded changeset does not change the model already" <>
+      "preset in the parent model"
+  end
+
+  # defp do_change(current, current) do
+  # TODO: What action do we set in that case?
+  # end
+
+  defp do_change(value, _current) do
+    %{Changeset.change(value) | action: :insert}
```

Comment:
> I don't know what to do with exactly same models. Should we even run callbacks in that case? I think it will be fine to treat it as an update.

Exactly same models can be an update where changes is an empty map.